### PR TITLE
make MAX_NUM_THREADS_PER_QUERY configurable for pinto server

### PR DESCRIPTION
**Pull request by samarth-gupta-traceable**
## Description
Fix for https://github.com/apache/incubator-pinot/issues/6374

Introduces a new system property that can be provided at startup time to configure number of threads per query for pinot server.  See referred issue for details. 

## Upgrade Notes

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options: yes
new system variable "pinot.server.maxThreadsPerQuery" introduced to configure max threads per query.

## Documentation
If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document

### CODE REVIEW

**Comment by siddharthteotia** [comment]
Why are we using the JVM system property based mechanism to set this as opposed to using pinot.server.* configurations which is how we set the server instance config?

Also I feel as opposed to making this as a server/instance level config, we should allow this as a table level config or at least both.  Will be useful for multi-tenancy and later on implementing query or table level priority. There is a queryConfig field in table config which is used to pass down the table level query timeout information. I think we can enhance QueryConfig to pass down this information as well.

**Comment by mcvsubbu** [comment]
> Why are we using the JVM system property based mechanism to set this as opposed to using pinot.server.* configurations which is how we set the server instance config?
> 
> Also I feel as opposed to making this as a server/instance level config, we should allow this as a table level config or at least both. Will be useful for multi-tenancy and later on implementing query or table level priority. There is a queryConfig field in table config which is used to pass down the table level query timeout information. I think we can enhance QueryConfig to pass down this information as well.

+1 to not using jvm property, and use what pinot does with other config items.

Number of threads to use does seem to be a per-system property, though. Even in a multi-tenant system, the table configs will have to carry a weighted value for num threads, which will get hard to provision across tables. I think having this as a per-host config is better.

**Comment by kishoreg** [comment]
> > Why are we using the JVM system property based mechanism to set this as opposed to using pinot.server.* configurations which is how we set the server instance config?
> > Also I feel as opposed to making this as a server/instance level config, we should allow this as a table level config or at least both. Will be useful for multi-tenancy and later on implementing query or table level priority. There is a queryConfig field in table config which is used to pass down the table level query timeout information. I think we can enhance QueryConfig to pass down this information as well.
> 
> +1 to not using jvm property, and use what pinot does with other config items.
> 
> Number of threads to use does seem to be a per-system property, though. Even in a multi-tenant system, the table configs will have to carry a weighted value for num threads, which will get hard to provision across tables. I think having this as a per-host config is better.

why per host? it's better to control this per table right?

**Comment by kishoreg** [comment]
> Why are we using the JVM system property based mechanism to set this as opposed to using pinot.server.* configurations which is how we set the server instance config?
> 
> Also I feel as opposed to making this as a server/instance level config, we should allow this as a table level config or at least both. Will be useful for multi-tenancy and later on implementing query or table level priority. There is a queryConfig field in table config which is used to pass down the table level query timeout information. I think we can enhance QueryConfig to pass down this information as well.

+1 on using per table config

**Comment by samarth-gupta-traceable** [comment]
@siddharthteotia  I have made changes to make this part of QueryConfig. However I am familiar with the code enough and not sure how to pass this along and use in `CombineOperatorUtils`  or other place in call hierarchy . 
Some pointer or code examples will be helpful. Thanks 

cc @kishoreg

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java [hunk 1]
```diff
@@ -236,7 +236,7 @@ public void testRefreshTableConfigAndQueryTimeout()
       throws Exception {
     // Set timeout as 5ms so that query will timeout
     TableConfig tableConfig = getOfflineTableConfig();
-    tableConfig.setQueryConfig(new QueryConfig(5L));
+    tableConfig.setQueryConfig(new QueryConfig(5L, 2));
```
Comment (mayankshriv) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java:239]: This seems like a change from current behavior. Any reason why we want to constrain the number of cores to be used to 2 for integration tests?

Comment (samarth-gupta-traceable) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java:239]: @mayankshriv  I am not that familiar with integration tests code. There is not specific reason to use value of 2 here. Pls suggest a reasonable value that can be used here. thanks

**Pull request closed without merging**